Tools/ardupilotwaf/cxx_checks.py

```diff
@@ -160,13 +160,38 @@ def check_librt(cfg, env):
 def check_package(cfg, env, libname):
     '''use pkg-config to look for an installed library that has a LIBNAME.pc file'''
     capsname = libname.upper()
+
+    # we don't want check_cfg() changing the global environment during
+    # this test, in case it fails in the 2nd link step
+    cfg.env.stash()
+
     cfg.check_cfg(package=libname, mandatory=False, global_define=True,
                   args=['--libs', '--cflags'], uselib_store=capsname)
 
-    env.LIB += cfg.env['LIB_%s' % capsname]
-    env.INCLUDES += cfg.env['INCLUDES_%s' % capsname]
-    env.CFLAGS += cfg.env['CFLAGS_%s' % capsname]
-    env.LIBPATH += cfg.env['LIBPATH_%s' % capsname]
+    # we need to also check that we can link against the lib. We may
+    # have a pc file for the package, but its the wrong
+    # architecture. This can happen as PKG_CONFIG_PATCH is not
+    # architecture specific
+    cfg.env.LIB += cfg.env['LIB_%s' % capsname]
+    cfg.env.INCLUDES += cfg.env['INCLUDES_%s' % capsname]
+    cfg.env.CFLAGS += cfg.env['CFLAGS_%s' % capsname]
+    cfg.env.LIBPATH += cfg.env['LIBPATH_%s' % capsname]
+
+    ret = cfg.check(
+        compiler='cxx',
+        fragment='''int main() { return 0; }''',
+        msg='Testing link with %s' % libname,
+        mandatory=False,
+        lib='dl'
+    )
+
+    cfg.env.revert()
```

Comment:
Sorry for butting in after such a long time away from the project. I'm trying to get back little by little. Instead of stashing the environment, I believe we could simply add `use=capsname` to the argument list for `cfg.check()`. That way we can add the flags only after a successful linkage. Here is an example:

```python
def options(opt):
    opt.load('compiler_c')

def configure(cfg):
    cfg.load('compiler_c')

    cfg.check_cfg(
        package='ncurses',
        args=['--libs', '--cflags'],
        uselib_store='NCURSES',
    )

    cfg.check(
        compiler='c',
        fragment='''int main() { return 0; }''',
        msg='Testing linking with ncurses',
        use='NCURSES',
    )

def build(bld):
    pass
```

Calling `waf configure -v` yields the following:
```
[0mSetting top to                           :[0m [32m/tmp/test[0m 
[0mSetting out to                           :[0m [32m/tmp/test/build[0m 
[0mChecking for 'gcc' (C compiler)          :[0m 22:57:07 [35mrunner[0m ['/usr/bin/gcc', '-dM', '-E', '-']
[32m/usr/bin/gcc[0m 
[0mChecking for program 'pkg-config'        :[0m [32m/usr/bin/pkg-config[0m 
[0mChecking for 'ncurses'                   :[0m 22:57:07 [35mrunner[0m ['/usr/bin/pkg-config', '--libs', '--cflags', 'ncurses']
[32myes[0m 
[0mTesting linking with ncurses             :[0m 22:57:07 [35mrunner[0m ['/usr/bin/gcc', '-DHAVE_NCURSES=1', '-D_GNU_SOURCE', '-D_DEFAULT_SOURCE', '../test.c', '-c', '-o/tmp/test/build/.conf_check_ebd1e6430b68bad08ed4a667c322d759/testbuild/test.c.1.o']
22:57:07 [35mrunner[0m ['/usr/bin/gcc', 'test.c.1.o', '-o/tmp/test/build/.conf_check_ebd1e6430b68bad08ed4a667c322d759/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic', '-lncursesw']
[32myes[0m 
'configure' finished successfully (0.213s)
```

Sorry about the control characters :-P